File: backend/routes/report.js
// backend/routes/report.js

const express = require('express');
const supabase = require('../config/supabase');
const protect = require('../middleware/authMiddleware');
const router = express.Router();

router.get('/', protect, async (req, res) => {
  try {
    // Get all subscriptions
    const { data: subscriptions } = await supabase
      .from('subscriptions')
      .select('*')
      .eq('user_id', req.user.userId);

    if (!subscriptions || subscriptions.length === 0) {
      return res.json({
        totalMonthly: 0,
        totalYearly: 0,
        byCategory: [],
        topSubscriptions: [],
        potentialSavings: 0,
        noRecentReceipt: [],
        monthlyTrend: []
      });
    }

    // ── Total monthly spend ──
    const totalMonthly = subscriptions.reduce(
      (sum, s) => sum + Number(s.amount), 0
    );

    // ── Group by category ──
    const categoryMap = {};
    subscriptions.forEach(sub => {
      const cat = sub.category || 'Other';
      if (!categoryMap[cat]) categoryMap[cat] = 0;
      categoryMap[cat] += Number(sub.amount);
    });

    const byCategory = Object.entries(categoryMap)
      .map(([name, amount]) => ({
        name,
        amount,
        percentage: Math.round((amount / totalMonthly) * 100)
      }))
      .sort((a, b) => b.amount - a.amount);

    // ── Top subscriptions ──
    const topSubscriptions = [...subscriptions]
      .sort((a, b) => b.amount - a.amount)
      .slice(0, 6)
      .map(s => ({
        id: s.id,
        name: s.service_name,
        monthly: s.amount,
        yearly: s.amount * 12,
        category: s.category,
        lastReceipt: s.last_receipt_date
      }));

    // ── Find subscriptions with no recent receipt ──
    const today = new Date();
    const noRecentReceipt = subscriptions.filter(sub => {
      if (!sub.last_receipt_date) return false;
      const lastReceipt = new Date(sub.last_receipt_date);
      const days = Math.floor(
        (today - lastReceipt) / (1000 * 60 * 60 * 24)
      );
      return days > 90;  // No receipt in 90+ days
    });

    const potentialSavings = noRecentReceipt.reduce(
      (sum, s) => sum + Number(s.amount), 0
    );

    // ── Monthly trend from receipts ──
    const sixMonthsAgo = new Date();
    sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);

    const { data: receipts } = await supabase
      .from('receipts')
      .select('amount, receipt_date')
      .eq('user_id', req.user.userId)
      .gte('receipt_date', sixMonthsAgo.toISOString().split('T')[0]);

    // Group receipts by month
    const monthlyMap = {};
    (receipts || []).forEach(r => {
      if (!r.receipt_date) return;
      const month = r.receipt_date.substring(0, 7); // YYYY-MM
      if (!monthlyMap[month]) monthlyMap[month] = 0;
      monthlyMap[month] += Number(r.amount);
    });

    const monthlyTrend = Object.entries(monthlyMap)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([month, amount]) => ({ month, amount }));

    res.json({
      totalMonthly,
      totalYearly: totalMonthly * 12,
      byCategory,
      topSubscriptions,
      potentialSavings,
      noRecentReceipt: noRecentReceipt.map(s => ({
        id: s.id,
        name: s.service_name,
        amount: s.amount,
        lastReceipt: s.last_receipt_date
      })),
      monthlyTrend
    });

  } catch (err) {
    console.error('Report error:', err);
    res.status(500).json({ error: 'Failed to generate report' });
  }
});

module.exports = router;

File: backend/routes/report.py
# backend/routes/report.py
# Spending report — port of routes/report.js

from fastapi import APIRouter, Depends, HTTPException
from datetime import datetime, timedelta
from math import floor

from config.database import supabase
from middleware.auth import get_current_user

router = APIRouter()


@router.get("/")
async def get_report(user: dict = Depends(get_current_user)):
    try:
        result = supabase.table("subscriptions") \
            .select("*") \
            .eq("user_id", user["userId"]) \
            .execute()

        subscriptions = result.data or []

        if not subscriptions:
            return {
                "totalMonthly": 0,
                "totalYearly": 0,
                "byCategory": [],
                "topSubscriptions": [],
                "potentialSavings": 0,
                "noRecentReceipt": [],
                "monthlyTrend": []
            }

        # Total monthly spend
        total_monthly = sum(float(s.get("amount", 0)) for s in subscriptions)

        # Group by category
        category_map = {}
        for sub in subscriptions:
            cat = sub.get("category", "Other")
            category_map[cat] = category_map.get(cat, 0) + float(sub.get("amount", 0))

        by_category = sorted([
            {
                "name": name,
                "amount": amount,
                "percentage": round((amount / total_monthly) * 100) if total_monthly > 0 else 0
            }
            for name, amount in category_map.items()
        ], key=lambda x: x["amount"], reverse=True)

        # Top subscriptions
        sorted_subs = sorted(subscriptions, key=lambda s: float(s.get("amount", 0)), reverse=True)
        top_subscriptions = [
            {
                "id": s["id"],
                "name": s["service_name"],
                "monthly": s["amount"],
                "yearly": float(s["amount"]) * 12,
                "category": s.get("category"),
                "lastReceipt": s.get("last_receipt_date")
            }
            for s in sorted_subs[:6]
        ]

        # No recent receipt (90+ days)
        today = datetime.utcnow()
        no_recent = []
        for sub in subscriptions:
            if sub.get("last_receipt_date"):
                last = datetime.fromisoformat(sub["last_receipt_date"])
                days = floor((today - last).total_seconds() / 86400)
                if days > 90:
                    no_recent.append(sub)

        potential_savings = sum(float(s.get("amount", 0)) for s in no_recent)

        # Monthly trend from receipts
        six_months_ago = today - timedelta(days=180)
        receipts_result = supabase.table("receipts") \
            .select("amount, receipt_date") \
            .eq("user_id", user["userId"]) \
            .gte("receipt_date", six_months_ago.strftime("%Y-%m-%d")) \
            .execute()

        monthly_map = {}
        for r in (receipts_result.data or []):
            if r.get("receipt_date"):
                month = r["receipt_date"][:7]  # YYYY-MM
                monthly_map[month] = monthly_map.get(month, 0) + float(r.get("amount", 0))

        monthly_trend = sorted([
            {"month": month, "amount": amount}
            for month, amount in monthly_map.items()
        ], key=lambda x: x["month"])

        return {
            "totalMonthly": total_monthly,
            "totalYearly": total_monthly * 12,
            "byCategory": by_category,
            "topSubscriptions": top_subscriptions,
            "potentialSavings": potential_savings,
            "noRecentReceipt": [
                {"id": s["id"], "name": s["service_name"], "amount": s["amount"], "lastReceipt": s.get("last_receipt_date")}
                for s in no_recent
            ],
            "monthlyTrend": monthly_trend
        }

    except Exception as err:
        print(f"Report error: {err}")
        raise HTTPException(status_code=500, detail="Failed to generate report")
